### PR TITLE
Filling in more Alonzo API extensions for the Tx and TxBody

### PR DESCRIPTION
**Pull request by dcoutts**

### CODE REVIEW

**Comment by dcoutts** [comment]
For the moment this extends #2642. It will need to be rebased once that is merged.

File: cardano-api/src/Cardano/Api/ProtocolParameters.hs [hunk 1]
```diff
@@ -892,6 +897,94 @@ fromShelleyPParamsUpdate
...
+    ShelleyBasedEraMary    -> toShelleyPParams pparams
+    ShelleyBasedEraAlonzo  -> toAlonzoPParams  pparams
+
+toShelleyPParams :: ProtocolParameters -> Shelley.PParams ledgerera
+toShelleyPParams pparams =
+   Shelley.PParams
+     { Shelley._protocolVersion = let (maj, minor) = protocolParamProtocolVersion pparams
+                                  in Shelley.ProtVer maj minor
+     , Shelley._d = Shelley.unitIntervalFromRational $ protocolParamDecentralization pparams
+     , Shelley._extraEntropy  = toShelleyNonce $ protocolParamExtraPraosEntropy pparams
```
Comment (newhoggy) [inline cardano-api/src/Cardano/Api/ProtocolParameters.hs:917]: There's a double space here before the `=`.

File: cardano-api/src/Cardano/Api/ProtocolParameters.hs [hunk 2]
```diff
@@ -892,6 +897,94 @@ fromShelleyPParamsUpdate
...
+     , Shelley._tau = Shelley.unitIntervalFromRational $ protocolParamTreasuryCut pparams
+     }
+
+toAlonzoPParams :: ProtocolParameters -> Alonzo.PParams ledgerera
+toAlonzoPParams pparams =
+   Alonzo.PParams
+     { Alonzo._protocolVersion = let (maj, minor) = protocolParamProtocolVersion pparams
+                                 in Alonzo.ProtVer maj minor
+     , Alonzo._d = Shelley.unitIntervalFromRational $ protocolParamDecentralization pparams
+     , Alonzo._extraEntropy  = toShelleyNonce $ protocolParamExtraPraosEntropy pparams
```
Comment (newhoggy) [inline cardano-api/src/Cardano/Api/ProtocolParameters.hs:940]: Also here.

**Review by newhoggy — APPROVED** [review]

File: cardano-api/src/Cardano/Api/ProtocolParameters.hs (hunk 2 above)
Comment (newhoggy) [reply cardano-api/src/Cardano/Api/ProtocolParameters.hs]: Fixed

File: cardano-api/src/Cardano/Api/TxBody.hs [hunk 3]
```diff
@@ -1173,6 +1186,7 @@ deserialiseShelleyBasedTxBody mkTxBody bs =
           (CBOR.runAnnotator txbody fbs)
           (map (`CBOR.runAnnotator` fbs) txscripts)
           (CBOR.runAnnotator <$> txmetadata <*> pure fbs)
+          (error "deserialiseShelleyBasedTxBody: Redeemer pointer map")
```
Comment (Jimbo4350) [inline cardano-api/src/Cardano/Api/TxBody.hs:1219]: Yep, I wasn't sure how we wanted to handle serialisation.

Comment (newhoggy) [reply cardano-api/src/Cardano/Api/TxBody.hs:1219]: This bit is causing test failures currently.

Comment (newhoggy) [reply cardano-api/src/Cardano/Api/TxBody.hs:1219]:
```
    prop_roundtrip_txbody_shelley_CBOR:                        FAIL
        ✗ prop_roundtrip_txbody_shelley_CBOR failed at test/Test/Cardano/Api/Typed/CBOR.hs:156:3
          after 1 test.
        
              ┏━━ test/Test/Cardano/Api/Typed/CBOR.hs ━━━
          152 ┃ roundtrip_CBOR
          153 ┃   :: (SerialiseAsCBOR a, Eq a, Show a)
          154 ┃   => AsType a -> Gen a -> Property
          155 ┃ roundtrip_CBOR typeProxy gen =
          156 ┃   H.property $ do
          157 ┃     val <- H.forAll gen
          158 ┃     H.tripping val serialiseToCBOR (deserialiseFromCBOR typeProxy)
              ┃     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
              ┃     │ ━━━ Exception (ErrorCall) ━━━
              ┃     │ deserialiseShelleyBasedTxBody: Redeemer pointer map
              ┃     │ CallStack (from HasCallStack):
              ┃     │   error, called at src/Cardano/Api/TxBody.hs:1219:12 in cardano-api-1.27.0-inplace:Cardano.Api.TxBody
        
          This failure can be reproduced by running:
          > recheck (Size 0) (Seed 16246837987535537937 6040485580684346801) prop_roundtrip_txbody_shelley_CBOR
        
      Use '--hedgehog-replay "Size 0 Seed 16246837987535537937 6040485580684346801"' to reproduce.
    prop_roundtrip_txbody_allegra_CBOR:                        FAIL
        ✗ prop_roundtrip_txbody_allegra_CBOR failed at test/Test/Cardano/Api/Typed/CBOR.hs:156:3
          after 1 test.
        
              ┏━━ test/Test/Cardano/Api/Typed/CBOR.hs ━━━
          152 ┃ roundtrip_CBOR
          153 ┃   :: (SerialiseAsCBOR a, Eq a, Show a)
          154 ┃   => AsType a -> Gen a -> Property
          155 ┃ roundtrip_CBOR typeProxy gen =
          156 ┃   H.property $ do
          157 ┃     val <- H.forAll gen
          158 ┃     H.tripping val serialiseToCBOR (deserialiseFromCBOR typeProxy)
              ┃     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
              ┃     │ ━━━ Exception (ErrorCall) ━━━
              ┃     │ deserialiseShelleyBasedTxBody: Redeemer pointer map
              ┃     │ CallStack (from HasCallStack):
              ┃     │   error, called at src/Cardano/Api/TxBody.hs:1219:12 in cardano-api-1.27.0-inplace:Cardano.Api.TxBody
        
          This failure can be reproduced by running:
          > recheck (Size 0) (Seed 5526808619565871689 17014309501333539577) prop_roundtrip_txbody_allegra_CBOR
        
      Use '--hedgehog-replay "Size 0 Seed 5526808619565871689 17014309501333539577"' to reproduce.
    prop_roundtrip_txbody_mary_CBOR:                           FAIL
        ✗ prop_roundtrip_txbody_mary_CBOR failed at test/Test/Cardano/Api/Typed/CBOR.hs:156:3
          after 1 test.
        
              ┏━━ test/Test/Cardano/Api/Typed/CBOR.hs ━━━
          152 ┃ roundtrip_CBOR
          153 ┃   :: (SerialiseAsCBOR a, Eq a, Show a)
          154 ┃   => AsType a -> Gen a -> Property
          155 ┃ roundtrip_CBOR typeProxy gen =
          156 ┃   H.property $ do
          157 ┃     val <- H.forAll gen
          158 ┃     H.tripping val serialiseToCBOR (deserialiseFromCBOR typeProxy)
              ┃     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
              ┃     │ ━━━ Exception (ErrorCall) ━━━
              ┃     │ deserialiseShelleyBasedTxBody: Redeemer pointer map
              ┃     │ CallStack (from HasCallStack):
              ┃     │   error, called at src/Cardano/Api/TxBody.hs:1219:12 in cardano-api-1.27.0-inplace:Cardano.Api.TxBody
        
          This failure can be reproduced by running:
          > recheck (Size 0) (Seed 2450485195492175484 8218421726347182459) prop_roundtrip_txbody_mary_CBOR
        
      Use '--hedgehog-replay "Size 0 Seed 2450485195492175484 8218421726347182459"' to reproduce.
```

File: cardano-api/src/Cardano/Api/TxBody.hs [hunk 4]
```diff
@@ -1930,6 +1959,179 @@ makeShelleyTransactionBody era@ShelleyBasedEraAlonzo
...
+  -> StrictMaybe (Alonzo.WitnessPPDataHash (Ledger.Crypto ledgerera))
+makeTxWitnessPPDataHash sbe pParams langs rMap =
+  Alonzo.hashWitnessPPData
+    (toLedgerPParams sbe pParams)
+    langs
+    (Alonzo.Redeemers rMap)
+
+
+
+-- TODO: Better name?
```
Comment (Jimbo4350) [inline cardano-api/src/Cardano/Api/TxBody.hs:2030]: Leftover TODO here.

File: cardano-api/src/Cardano/Api/TxBody.hs [hunk 5]
```diff
@@ -1930,6 +1959,179 @@ makeShelleyTransactionBody era@ShelleyBasedEraAlonzo
...
+   onCert sCred (StakeAddressDeregistrationCertificate target) = sCred == target
+   onCert sCred (StakeAddressDelegationCertificate target _) = sCred == target
+   onCert _ _ = False
+certifyingRedeemerPtr _ = []
+
+-- TODO: Alonzo - Definitely want a property test that the constructed redeemers point to the
+--correct thing in the final tx. Also wrap in an Either
+mintingRedeemerPtr
+  :: TxMintValue BuildTx era
+  -- BuildTxWith build (Map PolicyId (Witness WitCtxMint era))
```
Comment (Jimbo4350) [inline cardano-api/src/Cardano/Api/TxBody.hs:2074]: Leftover comment (from me :smile: )

**Review by Jimbo4350 — APPROVED** [review]
LGTM!

File: cardano-api/src/Cardano/Api/TxBody.hs (hunk 3 above)
Comment (Jimbo4350) [reply cardano-api/src/Cardano/Api/TxBody.hs:1219]: I'll push a fix

**Review by Jimbo4350 — APPROVED** [review]
LGTM!

**Pull request merged**